File: extractSIFTs.py
import cv2
import sys
import os

# Set wd to this file's location
os.chdir(os.path.dirname(os.path.abspath(sys.argv[0])))

# parameters
#gamma = 0.9 #contrastThreshold
delta = 8 #sampling step

# Loop through 2 datasets
for i in range(1,3):
    img_path = "VOC2005_" + str(i) + "/PNGImages/"
    desc_path = "VOC2005_" + str(i) + "/DescSIFTs/"
    if not os.path.exists(desc_path):
        os.makedirs(desc_path)
    kp_path = "VOC2005_" + str(i) + "/kpSIFTs/"
    if not os.path.exists(kp_path):
        os.makedirs(kp_path)

    # Get subdir list
    subdir = os.listdir(img_path)

    # Loop through subdir
    for j in range(0, len(subdir)):
        img_dir = img_path + subdir[j]
        desc_dir = desc_path + subdir[j]
        if not os.path.exists(desc_dir):
            os.makedirs(desc_dir)
        kp_dir = kp_path + subdir[j]
        if not os.path.exists(kp_dir):
            os.makedirs(kp_dir)

        # Get image list
        imgs = os.listdir(img_dir)

        # Loop through images
        for k in range(0, len(imgs)):
            name = imgs[k].split('.')[0]

            # Extract descriptors
            img = cv2.imread(img_dir + '/' + imgs[k])
            gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

            dense = cv2.FeatureDetector_create("Dense")
            dense.setInt('initXyStep', delta)
            kp = dense.detect(gray)

            sift = cv2.SIFT()
            #sift = cv2.SIFT(contrastThreshold=gamma)
            kp, desc = sift.compute(gray, kp)

            remove = []
            for i in range(0, len(desc)):
                if sum(desc[i]) == 0:
                    remove.append(i)

            for i in sorted(remove, reverse=True):
                del kp[i]

            keypoints = cv2.drawKeypoints(img, kp, img.copy())
            cv2.imwrite(kp_dir + '/' + imgs[k], keypoints)

            desc_out = open(desc_dir + '/' + name + '.txt', 'w')
            for l in range(0, len(desc)):
                for m in range(0, 128):
                    desc_out.write(str(desc[l][m]).ljust(10))
                desc_out.write('\n')
            desc_out.close()

File: extractSIFTs.m
%% Setup
% Change the current folder to the folder of this m-file.
path = '~/Dropbox/DM/ASL/Project/';
cd(path);

% Change this to VLfeat location
vl_dir = '~/Downloads/vlfeat-0.9.20' ;

% Setup vlfeat for matlab
run(strcat(vl_dir, '/toolbox/vl_setup'));

%% Extract features
delta = 8; % sampling step
gamma = 0.018; % contrast threshold

% Loop through 2 dataset
for num=1:2
    path = strcat('VOC2005_', int2str(num), '/');
    img_path = strcat(path, 'PNGImages/');
    
    desc_path = strcat(path, 'DescSIFTs/');
    if (exist(desc_path)==0)
        mkdir(desc_path);
    end
    
    kp_path = strcat(path, 'kpSIFTs/');
    if (exist(kp_path)==0)
        mkdir(kp_path);
    end
    
    % Get subdir list
    listdir = dir(img_path);
    
    % Loop through subdir
    for cat=3:(size(listdir,1))
        folder = listdir(cat).name;
        img_dir = strcat(img_path, folder, '/');
        
        desc_dir = strcat(desc_path, folder, '/');
        if (exist(desc_dir)==0)
            mkdir(desc_dir);
        end
    
        kp_dir = strcat(kp_path, folder, '/');
        if (exist(kp_dir)==0)
            mkdir(kp_dir);
        end
        
        % Get image list
        imgs = dir([img_dir '*.png']);
        
        % Loop through images
        for im=1:size(imgs,1)
            I = imread(strcat(img_dir,imgs(im).name)) ;
            
            % Extract dense descriptors
            I_single = im2single(I);
            [f,d] = vl_phow(I_single, 'ContrastThreshold', gamma, 'Step', delta) ;
            
            % Plot keypoints
            kp=[];
            for i=1:size(d,2)
                if(sum(d(:,i))~=0)
                    kp=[kp,f(1:2,i)];
                end
            end
            
            f = figure('visible','off');
            image(I)
            hold on
            scatter(kp(1,:), kp(2,:), 'red')
            hold off
            saveas(f, strcat(kp_dir,imgs(im).name(1:end-4)), 'png');
            
            % Save descriptors
            save(strcat(desc_dir, imgs(im).name(1:end-4), '.mat'), 'd');
            
        end
    end
    
end

